[SPARK-21122][CORE][YARN][WIP] Address starvation issues when dynamic allocation is enabled

cin opened the pull request:
## What changes were proposed in this pull request?

When dynamic resource allocation is enabled on a cluster, it’s currently possible for one application to consume all the cluster’s resources, effectively starving any other application trying to start. This is particularly painful in a notebook environment where notebooks may be idle for tens of minutes while the user is figuring out what to do next (or eating their lunch). Ideally the application should give resources back to the cluster when monitoring indicates other applications are pending.

## How was this patch tested?

Unit tests and manual testing. Unit tests are in `PreemptionPolicySuite.scala`. Manual testing.
TODO: add description of manual testing.

Commit:
starvation elimination changes

Commit:
clarified comments on forced removal edge case

frreiss commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -156,6 +157,11 @@ private[spark] class ExecutorAllocationManager(
   // Metric source for ExecutorAllocationManager to expose internal status to MetricsSystem.
   val executorAllocationManagerSource = new ExecutorAllocationManagerSource
 
+  // Preemption policy state
```
Can we make this a more descriptive comment? What does it mean when a bit in this set is TRUE vs. FALSE?

frreiss commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -156,6 +157,11 @@ private[spark] class ExecutorAllocationManager(
   val executorAllocationManagerSource = new ExecutorAllocationManagerSource
 
+  // Preemption policy state
+  private[spark] val preemptableExecutors = new mutable.HashSet[String]
```
Please describe the policy for accessing this data structure from different threads. Do threads need to acquire a lock on the parent ExecutorAllocationManager object before reading/writing this field?

frreiss replied:
Make it clearer that this field is owned by the preemption policy, or make it part of the policy object itself.

frreiss commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -597,6 +611,10 @@ private[spark] class ExecutorAllocationManager(
     removeTimes.remove(executorId)
   }
 
+  def preemptExecutors(pe: PreemptExecutors): Unit = synchronized {
```
There should be a ScalaDoc comment here because this method is intended to be called from other classes.

frreiss commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedClusterMessage.scala:
```diff
@@ -116,6 +116,12 @@ private[spark] object CoarseGrainedClusterMessages {
 
   case class KillExecutors(executorIds: Seq[String]) extends CoarseGrainedClusterMessage
 
+  case class PreemptExecutors(
```
What is the semantics of this message? Do the forcedToLeave and askedToLeave sets override any values previously passed via a PremptExecutors message, or do they add to the set from previous instances of this message? Are these semantics consistent across different resource managers?

frreiss commented on core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala:
```diff
@@ -285,6 +292,11 @@ private[spark] class ExecutorAllocationManager(
 
     updateAndSyncNumExecutorsTarget(now)
 
+    if (preemptableExecutors.nonEmpty) {
```
Instead of having the preemption policy modify private fields of this class, triggering this asynchronous removal, it would be better to have the preemption policy call the existing (albeit unused) removeExecutor() method in this class.

frreiss commented on core/src/main/scala/org/apache/spark/PreemptionPolicy.scala:
```diff
@@ -0,0 +1,121 @@
+ *                           EAM update. preemptExecutors adds to this set, but EAM acts
+ *                           on and clears the set
+ */
+abstract class PreemptionPolicy(
```
I think this is more of a pluggable callback than a policy.

frreiss commented on core/src/main/scala/org/apache/spark/PreemptionPolicy.scala:
```diff
@@ -0,0 +1,121 @@
+    val numExecsToPreempt = pe.forcedToLeave.size + pe.numRequestedContainers
+    val execsToPreempt = mutable.LinkedHashSet(pe.forcedToLeave.toSeq: _*)
+    if (pe.numRequestedContainers > 0) {
+      execsToPreempt ++= findPreemptableExecutors(pe.numRequestedContainers)
```
Please document (with a code comment) that YARN will accept alternate sacrificial executors in place of the executors that YARN asked you to preempt; and that what this code is doing is choosing a replacement sacrifice.

frreiss commented on core/src/main/scala/org/apache/spark/PreemptionPolicy.scala:
```diff
@@ -0,0 +1,121 @@
+   * NOTE: does not consider non-idle executors
+   */
+  protected[spark] def findPreemptableExecutors(numExecsToKill: Int): Seq[String] = {
+    val blockMaster = SparkEnv.get.blockManager.master
```
Recommend that you make this logic clearer by extracting a tuple (has cached data, idle time) for each executor, then sorting that list of tuples, then taking first k from the sorted list. Or perhaps just sort all executors on the number of bytes of cached data and kill the first k in that list.

frreiss commented on core/src/test/scala/org/apache/spark/PreemptionPolicySuite.scala:
```diff
@@ -0,0 +1,147 @@
+/*
```
Is it possible to have a stress test (i.e. process a large number of randomized executor creation and preemption messages and make sure nothing crashes) that is not enabled by default but is still checked in here?

frreiss reviewed (commented):
Looking good overall, recommend some cleanup of the logic to make the semantics of the various preemption operations and data structures clearer.

Commit:
commit after code review...need to address suggestions still but commiting to look at another branch

Commit:
preemption policy changes based on code review feedback

Commit:
added more scaladoc and comments based on code review feedback
tucked executorsToPreempt into the policy and added accessor methods

Commit:
too many brackets...

Commit:
fixing up unit tests

Commit:
Merge branch 'master' into SPARK-21122/starvation-elimination

Commit:
abstracted removal of executors within policy..debatable but isolates
synchronization to the policy so changes made to ExecutorAllocationManager
won't necessarily impact the preemption policy.

TODO: make policy optional

Commit:
added logic to make preemption policy optional.

TODO: look into cleaning up unit test changes to make this work

Commit:
cleaned up unit test logic around preemption policy changes
removed the number of requested containers to be preempted variable
  in the preemption message since no longer in use. also simplifies
  logic
